intro-HPC/ch_running_batch_jobs.tex

```diff
@@ -40,6 +40,331 @@ \chapter{Running batch jobs}
+disadvantages to this, so be aware of automatically loaded modules whenever
+things go wrong: they may have something to do with it!
+
+In fact, an easy way to check the components and version numbers of those
+components of a toolchain is to simply load the toolchain and then list the
+modules that are loaded.
+
+To unload a module, one can use the \lstinline|module unload| command. It works
+consistently with the load command, and reverses the latter's effect. However,
+the dependencies of the package are NOT automatically unloaded; the user shall
+unload the packages one by one. One can however unload automatically loaded
+modules manually, to debug some problem. When the \lstinline|secondexample| module is unloaded,
+only the following modules remain:
+
+\begin{prompt}
+%\shellcmd{module unload \emph{secondexample}}%
+%\shellcmd{module list}%
+Currently Loaded Modulefiles:
+  1) examplemodule/1.2.3-intel-2018a        3) intel/2018a
+  2) examplelib/1.2-intel-2014b
+\end{prompt}
+
+Notice that the version was not specified: the module system is sufficiently
+clever to figure out what the user intends. However, checking the list of
+currently loaded modules is always a good idea, just to make sure \ldots
+
+\subsection{Purging all modules}
+\label{subsec:purging-modules}
+
+In order to unload all modules at once, and hence be sure to start in a clean
+state, you can use:
+
+\begin{prompt}
+%\shellcmd{module purge}%
+\end{prompt}
+
+However, on some VSC clusters you may be left with a very empty list of available
+modules after executing \lstinline|module purge|. On those systems, \lstinline|module av| will show you a list of
+modules containing the name of a cluster or a particular feature of a section of
+the cluster, and loading the appropriate module will restore the module list
+applicable to that particular system.
+
+Modules need not be loaded one by one; the two \lstinline|module load| commands can be
+combined as follows:
+
+\begin{prompt}
+%\shellcmd{module load \emph{examplemodule/1.2.3-intel-2018a} \emph{secondexample/2.7-intel-2018a}}%
+\end{prompt}
+
+This will load the two modules as well as their dependencies.
+
+\subsection{Explicit version numbers}
+\label{subsec:explicit-version-numbers}
+
+As a rule, once a module has been installed on the cluster, the executables or
+libraries it comprises are never modified. This policy ensures that the user's
+programs will run consistently, at least if the user specifies a specific
+version. Failing to specify a version may result in unexpected behaviour.
+
+Consider the following example: the user decides to use the \lstinline|examplemodule| module
+and at that point in time, just a single version 1.2.3,
+compiled with Intel is installed on the cluster. The user loads the library
+using:
+
+\begin{prompt}
+%\shellcmd{module load \emph{examplemodule}}%
+\end{prompt}
+
+rather than
+
+\begin{prompt}
+%\shellcmd{module load \emph{examplemodule/1.2.3-intel-2018a}}%
```

Comment:
I would use a single `example` module here with 2 versions `1.2.3` and `4.5.6` (no `intel` part in it, that's irrelevant here)